test/polygon_reduce.cpp

```diff
@@ -0,0 +1,480 @@
+      return false;
+  }
+  return true;
+}
+
+std::pair<int, int> findIndex(const Polygons &polys, int i) {
+  int outer = 0;
+  while (i > polys[outer].size()) i -= polys[outer++].size();
+  return std::make_pair(outer, i);
+}
+
+void Dump(const Polygons &polys) {
+  for (auto poly : polys) {
+    std::cout << "polys.push_back({" << std::setprecision(9) << std::endl;
+    for (auto v : poly) {
+      std::cout << "    {" << v.x << ", " << v.y << "},  //" << std::endl;
+    }
+    std::cout << "});" << std::endl;
+  }
+  for (auto poly : polys) {
+    std::cout << "show(array([" << std::endl;
+    for (auto v : poly) {
+      std::cout << "  [" << v.x << ", " << v.y << "]," << std::endl;
+    }
+    std::cout << "]))" << std::endl;
+  }
+}
+
+void DumpTriangulation(const Polygons &polys) {
+  ExecutionParams oldParams = PolygonParams();
+  manifold::PolygonParams().processOverlaps = true;
+  auto result = Triangulate(polys);
+  PolygonParams() = oldParams;
+  for (auto &tri : result) {
+    auto x = findIndex(polys, tri.x);
+    auto y = findIndex(polys, tri.y);
+    auto z = findIndex(polys, tri.z);
+    printf("show(array([[%.7f, %.7f], [%.7f, %.7f], [%.7f, %.7f]]))\n",
+           polys[x.first][x.second].x, polys[x.first][x.second].y,
+           polys[y.first][y.second].x, polys[y.first][y.second].y,
+           polys[z.first][z.second].x, polys[z.first][z.second].y);
+  }
+}
+
+bool triangulationValid(Polygons &polys, std::vector<glm::ivec3> triangles) {
+  std::vector<std::pair<glm::vec2, glm::vec2>> edges(3 * triangles.size());
+  for (auto &tri : triangles) {
+    auto x = findIndex(polys, tri.x);
+    auto y = findIndex(polys, tri.y);
+    auto z = findIndex(polys, tri.z);
+    edges.push_back(
+        std::make_pair(polys[x.first][x.second], polys[y.first][y.second]));
+    edges.push_back(
+        std::make_pair(polys[y.first][y.second], polys[z.first][z.second]));
+    edges.push_back(
+        std::make_pair(polys[z.first][z.second], polys[x.first][x.second]));
+  }
+  return std::all_of(edges.begin(), edges.end(), [&](auto &p) {
+    return all_of(ExecutionPolicy::Par, edges.begin(), edges.end(),
+                  [&](auto &q) {
+                    return !intersect(p.first, p.second, q.first, q.second);
+                  });
+  });
+}
+
+// we are assuming polys is valid
+// we try to remove vertices from polys such that
+// 1. the updated polys is still valid (no overlapping edges)
+// 2. error in triangulation (either geometryErr or overlapping triangles)
+void simplify(Polygons &polys, float precision = -1) {
+  ExecutionParams oldParams = PolygonParams();
+  PolygonParams().intermediateChecks = true;
+  PolygonParams().processOverlaps = false;
+  PolygonParams().suppressErrors = true;
+
+  bool removedSomething = true;
+  while (removedSomething) {
+    removedSomething = false;
+    for (int i = 0; i < polys.size(); i++) {
+      for (int j = 0; j < polys[i].size(); j++) {
+        if (safeToRemove(polys, i, j)) {
```

Comment:
I might invert this and use `continue` to reduce indentation and increase readability.